ddcommon-ffi/src/array_queue.rs

```diff
@@ -0,0 +1,294 @@
+pub struct ArrayQueue {
+    // The actual type here should be *mut crossbeam_queue::ArrayQueue<*mut c_void>.
+    // However, cbindgen does not use the module name crossbeam_queue to generate the C header.
+    // So we use *mut c_void here and cast it to the correct type in the FFI functions.
+    inner: *mut c_void,
+    item_delete_fn: unsafe extern "C" fn(*mut c_void) -> c_void,
+}
+
+impl ArrayQueue {
+    pub fn new(
+        capacity: usize,
+        item_delete_fn: unsafe extern "C" fn(*mut c_void) -> c_void,
+    ) -> Self {
+        let internal_queue: crossbeam_queue::ArrayQueue<*mut c_void> =
+            crossbeam_queue::ArrayQueue::new(capacity);
+        let inner = Box::into_raw(Box::new(internal_queue)) as *mut c_void;
+        Self {
+            inner,
+            item_delete_fn,
+        }
+    }
+
+    pub fn take(&mut self) -> Option<Box<crossbeam_queue::ArrayQueue<*mut c_void>>> {
+        let raw = std::mem::replace(&mut self.inner, std::ptr::null_mut());
+
+        if raw.is_null() {
+            None
+        } else {
+            // # Safety: the raw pointer is not null and points to a valid memory location.
+            let queue =
+                unsafe { Box::from_raw(raw as *mut crossbeam_queue::ArrayQueue<*mut c_void>) };
+            while let Some(item) = queue.pop() {
+                // # Safety: the item is a valid memory location that can be deallocated by the
+                // item_delete_fn.
+                unsafe {
+                    (self.item_delete_fn)(item);
+                }
+            }
+            Some(queue)
+        }
+    }
+}
+
+impl<'a> ArrayQueue {
+    pub fn as_inner_ref(
+        queue: &'a mut ArrayQueue,
+    ) -> anyhow::Result<&'a crossbeam_queue::ArrayQueue<*mut c_void>> {
+        let inner = queue.inner;
+        if inner.is_null() {
+            anyhow::bail!("queue.inner is null");
+        }
+        // # Safety: the inner points to a valid memory location which is a
+        // crossbeam_queue::ArrayQueue<*mut c_void>.
+        Ok(unsafe { &*(inner as *mut crossbeam_queue::ArrayQueue<*mut c_void>) })
+    }
+}
+
+impl Drop for ArrayQueue {
+    fn drop(&mut self) {
+        drop(self.take())
+    }
+}
+
+#[allow(unused)]
+#[repr(C)]
+pub enum ArrayQueueNewResult {
+    Ok(ArrayQueue),
+    Err(Error),
+}
+
+/// Creates a new ArrayQueue with the given capacity and item_delete_fn.
+/// The item_delete_fn is called when an item is dropped from the queue.
+#[no_mangle]
+#[must_use]
+pub extern "C" fn ddog_ArrayQueue_new(
+    capacity: usize,
+    item_delete_fn: unsafe extern "C" fn(*mut c_void) -> c_void,
+) -> ArrayQueueNewResult {
+    let ffi_queue = ArrayQueue::new(capacity, item_delete_fn);
+    ArrayQueueNewResult::Ok(ffi_queue)
+}
+
+/// Drops the ArrayQueue.
+/// # Safety
+/// The pointer is null or points to a valid memory location allocated by array_queue_new.
+#[no_mangle]
+pub unsafe extern "C" fn ddog_ArrayQueue_drop(queue: &mut ArrayQueue) {
```

Comment:
do we want to alert on the case of double free?

Reply:
Added comment